### ISSUE TITLE
Issues with UnivariateKDE.evaluate

### ISSUE THREAD

**Issue by mwaskom** [statsmodels/statsmodels#2152]
I'm trying to understand what's causing some surprising behavior with the `KDEUnivariate` class.

``` python
import numpy as np
from statsmodels.nonparametric.kde import KDEUnivariate
x = np.random.randn(20)
kde = KDEUnivariate(x)
```

Specifically, the `kde.evaluate` method seems unable to reproduce the `kde.density` property:

``` python
kde.fit()
np.testing.assert_array_almost_equal(kde.density, kde.evaluate(kde.support))
```

```
AssertionError: 
Arrays are not almost equal to 6 decimals

(mismatch 99.21875%)
 x: array([ 0.001018,  0.000991,  0.000969,  0.000953,  0.000942,  0.000937,
        0.000937,  0.000943,  0.000953,  0.000969,  0.00099 ,  0.001015,
        0.001046,  0.001082,  0.001123,  0.00117 ,  0.001221,  0.001278,...
 y: array([  3.280454e-04,   3.543112e-04,   3.824256e-04,   4.124967e-04,
         4.446365e-04,   4.789620e-04,   5.155945e-04,   5.546600e-04,
         5.962887e-04,   6.406158e-04,   6.877806e-04,   7.379272e-04,...
```

This is a subtle difference:

``` python
import matplotlib.pyplot as plt
import seaborn
plt.plot(kde.support, kde.density, lw=5)
plt.plot(kde.support, kde.evaluate(kde.support), "--", lw=5)
```

![kde](https://cloud.githubusercontent.com/assets/315810/5519987/c83a0594-893a-11e4-9391-2362eaf359f1.png)

It seems to be because `evaluate()` is not picking up the `fft` option:

``` python
kde.fit(fft=False)
np.testing.assert_array_almost_equal(kde.density, kde.evaluate(kde.support))
```

Although, this fails a stricter test:

``` python
np.testing.assert_array_equal(kde.density, kde.evaluate(kde.support))
```

```
AssertionError: 
Arrays are not equal

(mismatch 68.0%)
 x: array([  3.280454e-04,   7.088984e-04,   1.425063e-03,   2.664943e-03,
         4.636106e-03,   7.503213e-03,   1.129811e-02,   1.583096e-02,
         2.065008e-02,   2.509707e-02,   2.847294e-02,   3.028022e-02,...
 y: array([  3.280454e-04,   7.088984e-04,   1.425063e-03,   2.664943e-03,
         4.636106e-03,   7.503213e-03,   1.129811e-02,   1.583096e-02,
         2.065008e-02,   2.509707e-02,   2.847294e-02,   3.028022e-02,...
```

More troubling is that trying to use `evaluate` with a non-gaussian kernel raises an exception:

``` python
kde.fit(kernel="tri", fft=False)
kde.evaluate(kde.support)
```

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-16-bb57467307a4> in <module>()
      1 kde.fit(kernel="tri", fft=False)
----> 2 kde.evaluate(kde.support)

/Users/mwaskom/anaconda/lib/python2.7/site-packages/statsmodels/nonparametric/kde.pyc in evaluate(self, point)
    256         """
    257         _checkisfit(self)
--> 258         return self.kernel.density(self.endog, point)
    259 
    260 

/Users/mwaskom/anaconda/lib/python2.7/site-packages/statsmodels/sandbox/nonparametric/kernels.pyc in density(self, xs, x)
    181         xs = np.asarray(xs)
    182         n = len(xs) # before inDomain?
--> 183         xs = self.inDomain( xs, xs, x )[0]
    184         if xs.ndim == 1:
    185             xs = xs[:,None]

/Users/mwaskom/anaconda/lib/python2.7/site-packages/statsmodels/sandbox/nonparametric/kernels.pyc in inDomain(self, xs, ys, x)
    168             return (xs, ys)
    169         else:
--> 170             filtered = filter(isInDomain, zip(xs, ys))
    171             if len(filtered) > 0:
    172                 xs, ys = zip(*filtered)

/Users/mwaskom/anaconda/lib/python2.7/site-packages/statsmodels/sandbox/nonparametric/kernels.pyc in isInDomain(xy)
    163             """Used for filter to check if point is in the domain"""
    164             u = (xy[0]-x)/self.h
--> 165             return u >= self.domain[0] and u <= self.domain[1]
    166 
    167         if self.domain is None:

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

**Comment by rgommers** [comment]
The difference between ``fft=False`` and ``fft=True`` is expected; you're estimating the density with a different algorithm hence small differences result. The plot shows that both methods are pretty close, so everything is OK there.

Second test: you can't compare floating point numbers for exact equality - this should pass when using ``assert_allclose`` or similar with a small ``rtol``.

The failure for ``array_like`` input to ``evaluate`` is known: https://github.com/statsmodels/statsmodels/blob/master/statsmodels/nonparametric/tests/test_kde.py#L37. Needs to be implemented still it looks like. There's a reason that those kernels are still in the sandbox (https://github.com/statsmodels/statsmodels/blob/master/statsmodels/sandbox/nonparametric/kernels.py) I'm afraid. Although ``UnivariateKDE`` really shouldn't be using stuff from the sandbox.

**Comment by mwaskom** [comment]
> The difference between fft=False and fft=True is expected; you're estimating the density with a different algorithm hence small differences result. The plot shows that both methods are pretty close, so everything is OK there.

Maybe I'm just confused about the api, but I expected using `kde.fit(fft=True)` and `kde.evaluate()` to be similar to `model.fit()` and `model.predict()` elsewhere in statsmodels. Is there any way other than `evaluate` to get an estimate of the density at specific points?

> Second test: you can't compare floating point numbers for exact equality - this should pass when using assert_allclose or similar with a small rtol.

I understand this point in general, but I raised it as an issue because I was surprised that `kde.evaluate(kde.support)` was not *exactly how* `kde.density` is created.

> There's a reason that those kernels are still in the sandbox 

Perhaps, but that sandbox status isn't at all clear to the caller, unless they inspect the source file names in the traceback.

**Comment by rgommers** [comment]
I don't think that it's easily possible to make ``kde.fit(fft=True)`` and ``kde.evaluate()`` to use the same algorithm, because ``kdensityfft`` evaluates points on a regular grid only. @jseabold correct me if I'm wrong.

And no, there's no other way than ``evaluate`` to get an estimate at specific points.

**Comment by rgommers** [comment]
Answer for point 2 is the same - ``fit()`` uses a uniform grid hence it cannot be re-used exactly the same by ``evaluate``.

**Comment by rgommers** [comment]
I agree, the docstring should be updated to indicate what works and doesn't work.

**Comment by mwaskom** [comment]
> Answer for point 2 is the same - fit() uses a uniform grid hence it cannot be re-used exactly the same by evaluate.

Sorry, still don't understand. Isn't `kde.support` the uniform grid that `fit()` is using? In that test I'm not passing arbitrary values, I'm passing exactly the values that should correspond with the stored `density` array.

**Comment by mwaskom** [comment]
This is additionally confusing because calling `evaluate` without having called `fit` raises a `ValueError` stating that `fit` must be called first, leading one to think that `evaluate` is going to use the results of what happens when you call `fit`.

**Comment by rgommers** [comment]
Well, ``evaluate`` should be able to handle any input. It cannot just switch to another algorithm when you pass it a regularly spaced input right?

Here is the algorithm that ``evaluate`` uses: https://github.com/statsmodels/statsmodels/blob/master/statsmodels/sandbox/nonparametric/kernels.py#L185

And this is what ``fit`` uses: https://github.com/statsmodels/statsmodels/blob/master/statsmodels/nonparametric/kde.py#L270

``fit`` assumes a regular grid and is therefore much more efficient.

**Comment by rgommers** [comment]
And yes, I fully agree that it's confusing. The docstring should very clearly explain why ``fit`` and ``evaluate`` are not the same and when to use which. Having to use ``fit`` before ``evaluate`` is also not necessary, that could be done under the hood.

**Comment by mwaskom** [comment]
> Well, evaluate should be able to handle any input. It cannot just switch to another algorithm when you pass it a regularly spaced input right?

My point is just that this is basically what's happening, it's silently switching to a different algorithm from what was used by `fit`. It sounds like using `fit(kde=True)` and `evaluate` aren't compatible, so perhaps `evaluate` should raise in this case.

**Comment by mwaskom** [comment]
Actually I should also say for my purposes (plotting) I do intent to pass `evaluate` a regular grid. I just want to have control over the extent of the grid in a way not currently afforded by the `clip` parameter. Though I suspect this issue might hit others looking to do more sophisticated things with the object.

**Comment by rgommers** [comment]
@mwaskom it's been a while since I looked at it in detail, but I think using ``KDEMultivariate`` will work with 1-D data and use the same algorithm for fit/evaluate (fit is in ``__init__`` and evaluate is ``pdf).

**Comment by jseabold** [comment]
I think these comments may be addressed by #1678 and #973. I'll see what I can do to get those in and make sure the status quo addresses the comments here.

**Comment by josef-pkt** [comment]
The main new issue is the refactoring in PR #1871 which needs at least several solid days of review.

I only paid partially attention to this issue so far. The main problem is that there are too many versions and possibilities. AFAIK, there is no way to use fft to `evaluate` at a new point unless we use interpolation or similar.

Padarn fixed several bugs and made several improvements in the kernel code. However, one of the main issues was a refactoring to clear up the different versions, which is partially incorporated or superseded by the new refactoring.  We should break backwards compatibility only once.

**Comment by PierreBdR** [comment]
For what it's worth, in the PR #1871 , there are two families of functions available on the fitted models: one compute the data on grids and use FFT/DCT, and the other do point evaluation, so without any transform. Hopefully, that will be less confusing.

**Comment by jseabold** [comment]
If you want to submit a PR that would be helpful. 

My main issue with #973 and #1678 is that cleaning it up going to be a bit of work, and I don't want to do it if these changes are going to be made elsewhere as well. For me, those PRs need a separation of bug fix, refactoring, and enhancement commits. Squashing of typo fixes. Useful commit messages. Removing the merging of master in to the branch, etc. Don't want to go through this if the changes aren't going to be made or just changed again. That said, I can probably finish that today, so I have a slight preference on doing that rather than punting another year on KDE changes, but let me look at your PR.

**Comment by josef-pkt** [comment]
@jseabold  if you have a day available for statsmodels, then reviewing kalman filter, statespace model and SARIMAX would be more important.
